### ISSUE TITLE
Something broke the connection to homegear-gateway for MAX! (again)

### ISSUE THREAD

**Issue by Flole998** [Homegear/Homegear#369]
Unfortunately the connection to homegear-max gateways is no longer possible. The gateway logs:

`Dropping connection to client 7: Error during TLS handshake (handshake took 5000 ms): Resource temporarily unavailable, try again.`

while the client/homegear logs

```
04/18/23 21:51:03.082 Module MAX: MAX! Homegear Gateway "Max": Warning: Connection to device closed. Trying to reconnect...
04/18/23 21:51:04.083 Module MAX: MAX! Homegear Gateway "Max": Info: Successfully connected.
04/18/23 21:51:09.083 Module MAX: MAX! Homegear Gateway "Max": Error in file PhysicalInterfaces/HomegearGateway.cpp line 158 in function void MAX::HomegearGateway::listen(): Connection closed (4).
```

For Homematic the connection is working.

@hfedcba Tagging you, hoping that you'll get an E-Mail notification this time.

**Comment by Flole998** [comment]
Now it's completely broken for Homematic and MAX! devices.... :(

For Homematic it looks like this and the gateway  only logs "new connection" messages:

```
04/26/23 01:24:57.285 Module HomeMatic BidCoS: HomeMatic Homegear Gateway "Homematic": Info: Successfully connected.
04/26/23 01:24:57.287 Module HomeMatic BidCoS: HomeMatic Homegear Gateway "Homematic": Error in file PhysicalInterfaces/HomegearGateway.cpp line 243 in function void BidCoS::HomegearGateway::listen(): Connection closed (4).
04/26/23 01:24:57.287 Module HomeMatic BidCoS: HomeMatic Homegear Gateway "Homematic": Warning: Connection to device closed. Trying to reconnect...
```

**Comment by hfedcba** [comment]
Hi @Flole998,

I'm going to debug and fix this this evening.

Cheers,

Sathya

**Comment by Flole998** [comment]
Thanks a lot!

**Comment by hfedcba** [comment]
Hi @Flole998,

I tested it with HomeMatic BidCoS and EnOcean - on my test system I can't reproduce the error.

Could you give me some more information? What system is Homegear Gateway and Homegear running on (Raspberry Pi?)? Which version do you have installed. Could you update (if not done yet) to the current nightly version of Homegear and Homegear Gateway and see if the error still persists?

Cheers,

Sathya

**Comment by Flole998** [comment]
Hi,

I'm running Bidcos and Max Gateways on a raspberry pi 3. Homegear itself is running on a odroid c2 with Armbian (so it's running the debian version). I'm always running the latest dev version, no luck there. I'll try to see if I can do some debugging myself. I believe there's a different issue for Bidcos and Max devices, one appeared before the other. Max might be possible for you to reproduce.

I'll try to grab some Wireshark traces to hopefully get some insights there.

Thanks for your help

**Comment by Flole998** [comment]
So for the MAX! issue the TCP connection is established and without a TLS Handshake immediately killed again by homegear-gateway. For Homematic the TLS Handshake is completed and Homegear-Gateway sends the first encrypted packet, then Homegear answers and then Homegear-Gateway kills the connection. I will try to get additional insights on what exactly is going on, but the whole libc1-net is without debug symbols so it's a little challenging to debug.

**Comment by hfedcba** [comment]
> For Homematic the TLS Handshake is completed and Homegear-Gateway sends the first encrypted packet, then Homegear answers and then Homegear-Gateway kills the connection.

ok - I currently can only test communication after the initial connect for EnOcean. With EnOcean it works on my test system.

> but the whole libc1-net is without debug symbols so it's a little challenging to debug.

I'm temporarily not stripping the debug symbols now: https://gitit.de/sensaru-public/c1/libc1-net/-/commit/61886ece1ded4dc1950ec0dd1b3a4f7c0b34069d

The unstripped binary is in the next nightly.

**Comment by Flole998** [comment]
I already digged myself through the assembly ;) 

I think I am seeing what's wrong: The variable `stop_server_` is at 0x1000, so it looks like there is some null-pointer somewhere.... Going further up through the variables in `TcpServer.h` there is `tcp_server_info_` at 0xe70.  So apparently the server is not correctly initialized?

**Comment by Flole998** [comment]
Also I ran the libc1-net through Coverity static analysis and it found a ressource leak  `server_info` is going out of scope and might not have been freed at https://gitit.de/sensaru-public/c1/libc1-net/-/blob/master/src/TcpClient.cpp#L299, which seems to be the case for non-failed connections. Basically the same thing is also happening at TcpSocket.cpp:370. In TcpSocket.cpp:259 (which is part of a for-loop) the pointer is overwritten, leaking the memory it potentially points to. Also it is complaining that for https://gitit.de/sensaru-public/c1/libc1-net/-/blob/master/src/TcpSocket.cpp#L61 the variable `port_` is not initialized.

I don't think that's the cause of the issue, but I think it should be addressed to avoid issues in the future.

I also opened some PRs and issues for other things Coverity noticed. Unfortunately nothing that explains what I saw in gdb.

**Comment by hfedcba** [comment]
Hi @Flole998 ,

thank you for analyzing this. There was another error: The connect loop did not exit when `connect()` returned success immediately.

> I think I am seeing what's wrong: The variable stop_server_ is at 0x1000, so it looks like there is some null-pointer somewhere.... 

Yes, indeed. You found the issue. Whenever a packet was received before the RPC server was started, it accessed the TcpServer object as a nullptr. It should be fixed in the next nightly.

Hopefully this was it...

Thank you so much!

**Comment by Flole998** [comment]
Unfortunately it looks like that wasn't the solution yet.  I just installed the new libc1-net (Revision 37) aswell as homegear gateway and libhomegear-base (revision 3663). I will check to see if it's still null or if there's another issue maybe.

**Comment by hfedcba** [comment]
I'll debug it with a CUL and HomeMatic BidCoS later as well.

**Comment by Flole998** [comment]
I just checked and there is still a null-pointer involved. I will try to see if I can figure out where/how that null-pointer gets there....

**Comment by Flole998** [comment]
I am looking at it with a disassembler right now and there is simply an instruction in `TcpServer::Listen` that overwrites `this`. I have absolutely no clue where that is coming from though (so which line maps to this instruction). It appears to be the last instruction before the loop, so `std::unordered_map<int32_t, PTcpClientData> clients;` could be the one. I think when `this` is passed to the new listen-thread something might be wrong there so it is overwritten later on.....

**Comment by hfedcba** [comment]
Thanks again for trying to debug this (btw. I fixed the double-free issue). I tried to reproduce the issue using a CUL on an amd64 system without success. I have no idea as well what causes the pointer to get invalidated. Next step for me is to debug this on an ARM system.

**Comment by Flole998** [comment]
I just had a brief look through all of the libc1-net packages available for debian bullseye and only the arm64 contains floating point operations there, based on that I would guess that it can only be reproduced on arm64. I think I might need to setup a build env myself and then remove code, compile, look at the assembly and see when that offending instruction goes away.

**Comment by Flole998** [comment]
I think I figured out some of the issues with Homegear-Max: When I do a diff on homematic-bidcos and max there are several differences:

- In the destructor there is a call to `_bl->threadManager.join(_initThread);`
- Huge difference in how C1Net is setup (`tls = true` vs `tls = _settings->ssl`, `.client_cert_file` and the key file aren't specified at all), this is most likely wrong for max, also the startListening call is in one case before and in one case after the C1Net init stuff, that (also in stopListening the "stopListening" call is at a different location, first vs. last command in that function)
- Line 309 in one case the `ex.what()` is wrapped in std::string and in the other case `ex` is `const`

I believe if those are fixed they should at least have the same issue.

And back to the homematic-bidcos issue: After about a second or so the destructor get's called which then closes the socket, so somehow the last reference get's lost. The last thing I am seeing is that the client is added to the epoll in line 824, and I believe then the reference counter of the socket goes to 0 shortly afterwards so it get's destroyed.

**Comment by Flole998** [comment]
I think I found another small bug (not the cause of this though): In TcpServer.cpp the events array has the size of max_connections and the size of max_connections is passed to epoll_wait. Shouldn't that be max_connections + 1 as the socket it's listening on is added aswell, so for 1 max connections it's waiting on 2 fds (the connection and the socket in case another client tries to connect)?

**Comment by Flole998** [comment]
Finally some success: I removed the `CollectGarbage(clients);` call and now it is working magically again. The condition `!client.second->socket` is the one that's causing the deletion, so the `socket` is missing there.

**Comment by Flole998** [comment]
@hfedcba Any idea what might cause this and how to fix it? Also could you have a look at what I mentioned in https://github.com/Homegear/Homegear/issues/369#issuecomment-1540454457 ?

**Comment by Flole998** [comment]
I've submitted a PR to fix homegear-max now, if that's merged only the issue with homegear-gateway/libc1-net is preventing it from working again.

**Comment by Flole998** [comment]
I have fixed it now apparently by doing these 2 changes in libc1-net:

```
@@ -602,7 +604,7 @@ void TcpServer::Listen(const int32_t thread_index) {
       if (tcp_server_info_.log_callback) tcp_server_info_.log_callback(1, "Critical: Could not create epoll file descriptor: " + std::string(strerror(errno)));
       return;
     }
-    struct epoll_event events[tcp_server_info_.max_connections];
+    struct epoll_event events[tcp_server_info_.max_connections + 1];

     int32_t nfds = 0;
     int32_t socket_descriptor = -1;
@@ -634,13 +636,13 @@ void TcpServer::Listen(const int32_t thread_index) {
         }

         do {
-          nfds = epoll_wait(epoll_fd, events, (int)tcp_server_info_.max_connections, 100);
+          nfds = epoll_wait(epoll_fd, events, (int)tcp_server_info_.max_connections + 1, 100);
         } while (nfds == -1 && errno == EINTR);

         if (nfds == 0) {
           static int64_t last_garbage_collection = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
           int64_t now = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
-          if (now - last_garbage_collection > 60000 || clients_.size() >= tcp_server_info_.max_connections) {
+          if (now - last_garbage_collection > 60000 || clients_.size() > tcp_server_info_.max_connections) {
             last_garbage_collection = now;
             CollectGarbage();
             CollectGarbage(clients);
```

The first change is because the epoll should look for changes in all clients plus on the server socket, so it needs to be max_connections + 1 there.

The second change is necessary because otherwise it would always do the garbage collection, which will kill the connection before it's actually established as it detects it as a dead/killed connection that should be cleaned.

**Comment by hfedcba** [comment]
Hi @Flole998,

sorry for looking at this so late again... You are absolutely right, the array size is wrong. This was not easy to find... Wow, that you found the issue.

I added your changes to libc1-net and merged the changes in Homegear MAX!

A big thank you!

Sathya
